File: plugins/robots/robotsGenerator/nxtOSEK/generators/simpleElements/waitForLightBlockGenerator.cpp
#include "waitForLightBlockGenerator.h"
#include "../../nxtOSEKRobotGenerator.h"
#include "../listsmartline.h"

using namespace robots::generator;

WaitForLightBlockGenerator::WaitForLightBlockGenerator()
{
}

void WaitForLightBlockGenerator::addInitAndTerminateCode(NxtOSEKRobotGenerator *nxtGen
			, QString const &port, qReal::Id const elementId)
{
	QString const initCode = "ecrobot_set_light_sensor_active(" + port + ");";
	if (!ListSmartLine::isContains(nxtGen->initCode(), initCode)) {
		QString const terminateCode = "ecrobot_set_light_sensor_inactive(" + port + ");";
		nxtGen->initCode().append(SmartLine(initCode, elementId));
		nxtGen->terminateCode().append(SmartLine(terminateCode, elementId));
	}
}

QList<SmartLine> WaitForLightBlockGenerator::convertElementIntoDirectCommand(NxtOSEKRobotGenerator *nxtGen
		, qReal::Id const elementId, qReal::Id const logicElementId)
{
	QList<SmartLine> result;

	QString const port = "NXT_PORT_S" + nxtGen->api()->stringProperty(logicElementId, "Port");

	QString const percents = nxtGen->api()->stringProperty(logicElementId, "Percents");
	QString const inequalitySign = transformSign(QString(nxtGen->api()->stringProperty(logicElementId
			, "Sign").toUtf8()));

	QString const condition = inequalitySign + " " + percents;

	result.append(SmartLine("while (!(ecrobot_get_light_sensor(" + port
			+ ") " + condition + "))", elementId));
	result.append(SmartLine("{", elementId));
	result.append(SmartLine("}", elementId));

	addInitAndTerminateCode(nxtGen, port, elementId);
	return result;
}

File: plugins/visualInterpreter/examples/robotsCodeGeneration/reactionsStorage/WaitForLightBlockGenerator.py
# Application condition
waitFor.id == max_used_id and not cur_node_is_processed

# Reaction
port = "NXT_PORT_S" + waitFor.Port
condition = convertCondition(waitFor.Sign) + " " + waitFor.Percents

wait_for_light_code = "while (!(ecrobot_get_light_sensor(" + port + ") " + condition + ")) {}\n"
wait_init_code = "ecrobot_set_light_sensor_active(" + port + ");\n"
wait_terminate_code = "ecrobot_set_light_sensor_inactive(" + port + ");\n"

if wait_init_code not in init_code:
  init_code.append(wait_init_code)
  terminate_code.append(wait_terminate_code)

code.append([wait_for_light_code])
id_to_pos_in_code[waitFor.id] = len(code) - 1

cur_node_is_processed = True
